src/Workspaces/Core/Portable/Workspace/Solution/Document.cs

```diff
@@ -276,15 +276,18 @@ public bool TryGetSemanticModel(out SemanticModel semanticModel)
                     return result;
                 }
 
-                // it looks like someone has set it. try to reuse same semantic model
-                if (original.TryGetTarget(out semanticModel))
+                // It looks like someone has set it. Try to reuse same semantic model, or assign the new model if that
+                // fails. The lock is required since there is no compare-and-set primitive for WeakReference<T>.
+                lock (original)
                 {
-                    return semanticModel;
-                }
+                    if (original.TryGetTarget(out semanticModel))
+                    {
+                        return semanticModel;
+                    }
 
-                // it looks like cache is gone. reset the cache.
-                original.SetTarget(result);
-                return result;
+                    original.SetTarget(result);
```

Comment:
Looks like we have 5 other calls to `SetTarget` - should we be worried about race conditions in any of them?

https://github.com/dotnet/roslyn/search?utf8=%E2%9C%93&q=SetTarget&type=

Reply:
it depends on whether weak reference is purely for caching or to preserve observable reference equality.

now I am thinking more I am not sure whether we ever guaranteed reference equality for semantic model.

we do so for compilation/symbols (original def and etc but not local, generic instances and etc) and syntaxtree/nodes. but we created new semantic model for Compilation.GetSemanticModel 

so, not sure why document.GetSemanticModel suddenly tries to guarantee that. otherwise, it is purely cache.

symbol equality is tied to compilation and semantic model. same goes for node.

anything we want that reference equality preserved we use AsyncLazy or SourceValue. the fact that SemanticModel suddently not use it seems kind of tip that we never supported reference equality and weak reference is there solely for perf reason.

otherwise, we could change GetSemanticModelAsync implementation to use AsyncLazy like others?

Reply:
> now I am thinking more I am not sure whether we ever guaranteed reference equality for semantic model.

From my recollection, this was intended to be a contract of the Workspace.  So much so that we put these sorts of assets in our features.  

> otherwise, we could change GetSemanticModelAsync implementation to use AsyncLazy like others?

I'm not sure if that would work.  AsyncLazy can say that it does not want to cache a result (which is what we'd want to prevent unbounded memory usage).  But if you don't ask to cache results, there's nothing (afaict) that will ensure it will return the existing value if you ask it for it's value multiple times (while someone is holding onto a previous result).

Reply:
Note: we could consider CachedWeakValueSource.  I think the purpose of it is to be able to compute values on demand, not hold them alive, but return the same value each time as long as the initial value is alive.

Note: looking at the impl, it uses double-checked locking.  So i already don't trust it.  But, at least it seems like this is the problem space it's trying to solve. :)

Reply:
➡️ There are two cases I couldn't rule out, and filed #25881 for them.